### PR TITLE
Separate form values from schema'd settings.

### LINKED ISSUE #426: Separate form value state from resultant settings schema'd object.
Having the form input values and the resultant settings object be one in the same causes issues in the UX for the settings form. A good example is number-type input (see: https://github.com/Automattic/woocommerce-connect-client/issues/188)

This issue should include the following:
- [x] Change `settings` state to be "form values"
- [x] Add method to convert "form values" into a settings schema object (coerce values, etc)
- [x] https://github.com/Automattic/woocommerce-connect-client/issues/397

### PR DESCRIPTION
**Pull request by jeffstieler**
Fixes #426 and #397.

To test: (basically kick the tires and make sure all functionality works as before, with the exception of)
* Make sure that service adjustment will accept an empty input
* Make sure that service adjustment will accept the value `1.`

### CODE REVIEW

File: client/components/wcc-settings-form/settings-item.js [hunk 1]
```diff
@@ -23,18 +22,19 @@ const SettingsItem = ( {
...
-	const updateSubSubValue = ( key, subKey, val ) => settingsActions.updateSettingsObjectSubField( id, key, subKey, val );
-	const removeArrayItem = ( idx ) => settingsActions.removeSettingsArrayFieldItem( id, idx );
+	const updateSubValue = ( key, val ) => settingsActions.updateSettingsField( [ id, key ], val );
+	const removeArrayItem = ( idx ) => settingsActions.removeSettingsField( [ id, idx ] );
 	const savePackage = ( packageData ) => packagesActions.savePackage( id, packageData );
 	const fieldRequired = ( -1 !== schema.required.indexOf( id ) );
-	const fieldValue = settings[ id ];
+	const fieldValue = form.settings[ id ];
 	const fieldSchema = schema.properties[ id ];
+	const fieldType = layout.type || fieldSchema.type || '';
```
Comment (DanReyLop) [inline client/components/wcc-settings-form/settings-item.js]: The type should always come from the `layout`, the `type` field in the layout is mandatory (or it should be).

Comment (jeffstieler) [inline client/components/wcc-settings-form/settings-item.js]: I'm not sure that's enforced yet - so I was going for being "liberal" with what we accept.

File: client/lib/utils/coerce-values.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * Retrieve a field's schema, handling referenced schema definitions if need be.
+ *
+ * @param {Object} fieldSchema - Schema object for field.
+ * @param {Object} definitions - Common definitions.
+ * @returns {Object} - Schema object for field, potentially pulled from $ref's definition.
+ */
+const getFieldSchema = ( fieldSchema, definitions ) => {
+	if ( fieldSchema.hasOwnProperty( '$ref' ) ) {
+		const definitionKey = fieldSchema.$ref.match( /^#\/definitions\/(.+)/ );
```
Comment (DanReyLop) [inline client/lib/utils/coerce-values.js]: I think `fieldSchema.$ref.split('#/definitions/')[1]` would be cleaner (but yet again, I hate RegExps too much, so feel free to ignore me)

File: client/lib/utils/coerce-values.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Retrieve a field's schema, handling referenced schema definitions if need be.
+ *
+ * @param {Object} fieldSchema - Schema object for field.
+ * @param {Object} definitions - Common definitions.
+ * @returns {Object} - Schema object for field, potentially pulled from $ref's definition.
+ */
+const getFieldSchema = ( fieldSchema, definitions ) => {
+	if ( fieldSchema.hasOwnProperty( '$ref' ) ) {
```
Comment (DanReyLop) [inline client/lib/utils/coerce-values.js]: This `hasOwnProperty` thing always seemed overly-verbose to me. Wouldn't it be easier to just check if `fieldSchema.$ref` is `truthy`? Same thing applies to your other uses of `hasOwnProperty`.

File: client/lib/utils/coerce-values.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	return value;
+};
+
+/**
+ * Coerce a form values object using the provided schema.
+ *
+ * @param {Object} schema - Schema containing type declarations.
+ * @param {Object} values - Form values.
+ * @param {Object} [definitions=null] - Optional. Schema definitions, parsed from schema if omitted.
```
Comment (DanReyLop) [inline client/lib/utils/coerce-values.js]: Is there any use case for this 3rd argument? If not, it's better to just remove it and add it later if needed.

File: client/state/form/settings/test/actions.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		const action = updateSettingsField( path, value );
+
+		expect( action ).to.eql( {
+			type: UPDATE_SETTINGS_FIELD,
+			path: 'testField',
+			value: 'testValue',
+		} );
+	} );
+
+	it( '#removeeSettingsField()', () => {
```
Comment (DanReyLop) [inline client/state/form/settings/test/actions.js]: Typo in `removee`

File: package.json [hunk 6]
```diff
@@ -84,6 +85,7 @@
     "redux-thunk": "^2.1.0",
     "reselect": "^2.5.1",
     "sprintf-js": "1.0.3",
+    "type-coerce": "^0.1.0",
```
Comment (DanReyLop) [inline package.json]: Not sure if I trust this library. Looking at the repo: https://github.com/wacky6/type-coerce/blob/master/lib/type-coerce.js it seems that it doesn't have any stars, there was only one commit, with no issues or PRs... the code is simple enough, but this is rougly the equivalent of copying and pasting from StackOverflow. I'd suggest looking for another library, or implementing it ourselves, or reviewing the `type-coerce` library as if it was part of this PR.

**Comment by DanReyLop** [comment]
The adjustment input works as expected, but the Fallback Rate input still has the previous behaviour (empty value is converted to 0 on blur).

Also, everything :boom: in the Self-Help page.

File: client/lib/utils/coerce-values.js (hunk 4 above)
Comment (jeffstieler) [inline client/lib/utils/coerce-values.js]: It's used on lines 74 and 76 below. It's essentially a mechanism to pass down the `definitions` in recursive calls to `coerceFormValues()`.

File: client/lib/utils/coerce-values.js (hunk 2 above)
Comment (jeffstieler) [inline client/lib/utils/coerce-values.js]: `fieldSchema.$ref.split('#/definitions/')[1]` can fail with a malformed `$ref` value. I'm going to pass on this for now.

**Comment by jeffstieler** [comment]
> Also, everything 💥 in the Self-Help page.

Fixed!

File: package.json (hunk 6 above)
Comment (jeffstieler) [inline package.json]: I agree, see https://github.com/Automattic/woocommerce-connect-client/pull/430/commits/1b18857558bafd04e2f59f461cf560a8e30faa58

**Comment by DanReyLop** [comment]
:shipit: 

If you are not going to remove the `NumberInput` component in this PR, better write an issue so we don't forget. This behaviour is better (less "hacky") that the `NumberInput` one.

**Pull request merged**